test/sequence/token_refresh_sequence_test.rb

```diff
@@ -7,82 +7,162 @@
+    response_headers.delete(:pragma) if failure_mode == :pragma_off
+
+    exchange_response_json = exchange_response.to_json
+    exchange_response_json = '<bad>' if failure_mode == :bad_json_response
+
+    stub_request(:post, @instance.oauth_token_endpoint)
+      .with(headers: headers,
+            body: body)
+      .to_return(status: failure_mode == :requires_scope ? 400 : 200,
+                 body: exchange_response_json,
+                 headers: response_headers)
+
+    stub_request(:post, @instance.oauth_token_endpoint)
+      .with(headers: headers,
+            body: body_with_scope)
+      .to_return(status: 200,
+                 body: exchange_response_json,
+                 headers: response_headers)
+
+    # To test rejection of invalid client_id for public client
     stub_request(:post, @instance.oauth_token_endpoint)
       .with(body: /INVALID/,
             headers: { 'Content-Type' => 'application/x-www-form-urlencoded' })
       .to_return(status: 401)
 
+    # To test rejection of invalid client_id for confidential client
+    auth_header = "Basic #{Base64.strict_encode64('INVALID_CLIENT_ID:' + @confidential_client_secret)}"
+    stub_request(:post, @instance.oauth_token_endpoint)
+      .with(headers: { 'Content-Type' => 'application/x-www-form-urlencoded',
+                       'Authorization' => auth_header })
+      .to_return(status: 401)
+  end
+
+  def all_pass
+    setup_mocks
     sequence_result = @sequence.start
 
     assert sequence_result.pass?, 'The sequence should be marked as pass.'
   end
 
   def test_all_pass_confidential_client
-    @instance.client_secret = SecureRandom.uuid
+    @instance.client_secret = @confidential_client_secret
     @instance.confidential_client = true
-    all_pass(true)
+    all_pass
   end
 
   def test_all_pass_public_client
     @instance.client_secret = nil
-    @instance.confidential_client = nil
-    all_pass(false)
+    @instance.confidential_client = false
+    all_pass
+  end
+
+  def test_fail_if_bad_token_type
+    setup_mocks(:bad_token_type)
+
+    sequence_result = @sequence.start
+    assert sequence_result.fail?
+  end
+
+  def test_fail_if_cache_control_off
+    setup_mocks(:cache_control_off)
+
+    sequence_result = @sequence.start
+    assert sequence_result.fail?
+  end
+
+  def test_fail_if_pragma_off
+    setup_mocks(:pragma_off)
+
+    sequence_result = @sequence.start
+    assert sequence_result.fail?
+  end
+
+  def test_fail_if_no_scope_returned
+    setup_mocks(:no_scope)
+
+    sequence_result = @sequence.start
+    assert sequence_result.fail?
+  end
+
+  def test_fail_if_no_access_token
+    setup_mocks(:no_access_token)
+
+    sequence_result = @sequence.start
+    assert sequence_result.fail?
+  end
+
+  def test_fail_if_bad_json_response
+    setup_mocks(:bad_json_response)
+
+    sequence_result = @sequence.start
+    assert sequence_result.fail?
+  end
+
+  def test_fail_if_scope_must_be_in_payload
```

Comment:
Do we need a corresponding test for if the scope must be excluded from the payload?

Reply:
Added another test.